C++/CLI section (#1)
AaronRobinsonMSFT/interop-doc: doc/interop-doc.md

These two concepts are important in interop &ndash; especially when involving the C# language. The two are closely related but are not equivalent and can cause confusion when thinking about interop. Both are defined in the [Terminology](#term) section but examples and additional clarity are warranted.

[Blittable types][doc_blittable] are those types that have the same layout in memory in both managed and native environments. Since blittable isn't defined in the [ECMA-335][spec_ecma335] specification, runtime implementations are free to optimize cases where types aren't technically blittable but can be treated as such under certain circumstances.

[Unmanaged types][doc_unmanaged_types] are defined in the C# specification and are not related to blittability at all.

> Consider the following .NET types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`. Every one of these types has the same well defined size and representation in both a managed and native environment - so they are blittable. All of them are also defined as C# unmanaged types.

Reviewer: I think the `a`->`the same` switch makes this read as saying the types all have the same size and representation as each other.

Maybe `a ... representation that is the same in both ...`?

> There are however unmanaged types that are not blittable. The .NET types `bool`, `char`, and enumerations are unmanaged types, but are not considered blittable. The reason for this is partially historical and partially based on ECMA-335. The `bool` is specified as a single byte in .NET, but historically is marshalled as an `int` to align with the Windows' `BOOL` type. The `char` is specified to be a two-byte 16-bit value, but the native mapping for a character in C/C++ could be either [`char`](https://en.cppreference.com/w/cpp/keyword/char) or [`wchar_t`](https://en.cppreference.com/w/cpp/keyword/wchar_t), which are 1 byte and 2 bytes respectively on the Windows platform. The `wchar_t` is particularly difficult given it is 4-bytes on some non-Windows platforms. Finally, the ECMA-335 specification in II.14.3 defines the underlying type of an enumeration may be any numeric type or a `bool` or `char`. This means that an enumeration can't always be considered blittable.

Reviewer: No reference link for `BOOL`, I take it?

Author: Done.

> There are however unmanaged types that are not blittable. The .NET types `bool`, `char`, and enumerations are not considered blittable but are unmanaged types. The reason for this is partially historical and partially based on ECMA-335. The `bool` is specified as a single byte in .NET but historically is marshalled as an `int` to align with the Windows' `BOOL` type. The `char` is specified to be a two byte 16-bit value, but the native mapping for a character in C/C++ could be either [`char`](https://en.cppreference.com/w/cpp/keyword/char) or [`wchar_t`](https://en.cppreference.com/w/cpp/keyword/wchar_t), which are 1 byte and 2 bytes respectively on the Windows platform. The `wchar_t` is particularly difficult given it is 4-bytes on some non-Windows platforms. Finally, the ECMA-335 specification in II.14.3 defines the underlying type of an enumeration may be any numeric type or a `bool` or `char`. This means that an enumeration can't always be considered blittable.

Reviewer: ```suggestion
There are however unmanaged types that are not blittable. The .NET types `bool`, `char`, and enumerations are unmanaged types, but are not considered blittable. The reason for this is partially historical and partially based on ECMA-335. The `bool` is specified as a single byte in .NET, but historically is marshalled as an `int` to align with the Windows' `BOOL` type. The `char` is specified to be a two-byte 16-bit value, but the native mapping for a character in C/C++ could be either [`char`](https://en.cppreference.com/w/cpp/keyword/char) or [`wchar_t`](https://en.cppreference.com/w/cpp/keyword/wchar_t), which are 1 byte and 2 bytes respectively on the Windows platform. The `wchar_t` is particularly difficult given it is 4-bytes on some non-Windows platforms. Finally, the ECMA-335 specification in II.14.3 defines the underlying type of an enumeration may be any numeric type or a `bool` or `char`. This means that an enumeration can't always be considered blittable.
```

Reviewer: Do we have a link for `BOOL`?

> The opposite case is also possible &ndash; blittable but not unmanaged. Single dimension arrays of blittables types are considered blittable by the interop system but are not unmanaged types. Considering single dimension arrays of blittable types as blittable is an example of a runtime implementation optimization.

Reviewer: ```suggestion
The opposite case is also possible &ndash; blittable but not unmanaged. Single dimension arrays of blittable primitive types are considered blittable by the interop system but are not unmanaged types. Considering single dimension arrays of blittable primitive types as blittable is an example of a runtime implementation optimization.
```

## C++/CLI <a name="cppcli"></a>

> Support for writing C++ to run in a .NET runtime is accomplished using a language called C++/CLI and specified in [ECMA-372][spec_ecma372]. At present, the [C++/CLI language][doc_cppcli] is only supported by the VC++ compiler and limited exclusively to the Windows platform. One of the goals of C++/CLI was to make interoperability between native and managed code seamless and sort of magic, that is why it is commonly referred to as IJW or "It-Just-Works". C++/CLI's goal of making interop simpler is laudable but given its design comes with non-trivial costs in terms of implementation and a deeper understanding of interop as opposed performing it from within a language like C#.

Reviewer: ```suggestion
Support for writing C++ to run in a .NET runtime is accomplished using a language called C++/CLI as specified in [ECMA-372][spec_ecma372]. At present, the [C++/CLI language][doc_cppcli] is only supported by the VC++ compiler and limited exclusively to the Windows platform. One of the goals of C++/CLI was to make interoperability between native and managed code seamless and sort of magic &ndash; that is why it is commonly referred to as IJW or "It-Just-Works". C++/CLI's goal of making interop simpler is laudable, but given its design, it comes with non-trivial costs in terms of implementation and requires a deeper understanding of interop as opposed performing it from within a language like C#.
```

There are 4 modes of operation for a C++/CLI compiled assembly. The documentation for these modes is relatively vague so additional details are added here. The VC++ compiler will compile a C++/CLI assembly when passed the [`/clr`](https://docs.microsoft.com/cpp/build/reference/clr-common-language-runtime-compilation) flag.

> Comprehensive official documentation for C++/CLI can be found [here][doc_cppcli].

Reviewer: Maybe move this above the previous paragraph? Seems like an intruder between the above sentences talking about the modes / details being added here and the details. (Also, I found it funny that it went from documentation for the modes being vague to pointing at comprehensive official documentation)

|Mode|Operation|
|-|-|
|_none_| The default mode. The resulting assembly may contain both native and managed code. The assembly is linked against `mscoree.lib` and only runs on .NET Framework. |
|`pure`| **Deprecated**. The resulting assembly contains only managed code. All native data types are converted to pure managed types. The assembly is linked against `mscoree.lib` and only runs on .NET Framework. |
|`safe`| **Deprecated**. The resulting assembly may contain both native and managed code, but no native memory dereferences are permitted in the native code. The intent of this build is to produce verifiable code (that is, no `unsafe` code). The assembly is linked against `mscoree.lib` and only runs on .NET Framework.|
|`NetCore`| This mode is identical to the default but links against `ijwhost.lib` and only runs on .NET Core 3.1 or .NET 5+. |

> If either of the two deprecated modes are desired it is strongly recommended to write in or port code to a pure .NET language (for example, C#).

Reviewer: ```suggestion
If either of the two deprecated modes are desired, it is strongly recommended to write in or port code to a pure .NET language (for example, C#).
```

### C++ language extensions <a name="cppcli_cpplangext"></a>

> The C++/CLI language is an  extension of C++ that attempts to express .NET interop [Concepts](#concepts) in C++. The most important being memory ownership and levels of indirection.

Reviewer: Either `&ndash;` instead of separate sentences or make the second part a complete sentence.

> The C++/CLI language is an  extension of C++ that attempts to express .NET interop [Concepts](#concepts) in C++ &ndash; the most important being memory ownership and levels of indirection.

Reviewer: ```suggestion
The C++/CLI language is an extension of C++ that attempts to express .NET interop [Concepts](#concepts) in C++ &ndash; the most important being memory ownership and levels of indirection.
```

**`&`** &ndash; A native reference to native memory. This is the [C++ construct](https://en.cppreference.com/w/cpp/language/reference) and has the same semantics. This should not be confused with the `&` defined above when talking about managed references in IL &ndash; a completely different language.

**`*`** &ndash; A native pointer to native memory. This is the [C++ construct](https://en.cppreference.com/w/cpp/language/pointer) and has the same semantics.

**`%`** &ndash; A managed reference, also called a ["tracking reference"](https://docs.microsoft.com/cpp/dotnet/how-to-use-tracking-references-in-cpp-cli), to managed memory. This is a new construct and provides symmetry with the native `&` construct but for managed.

> **`^`** &ndash; A managed pointer to managed memory. This is a new construct and provides symmetry with the native `*` construct but for managed.

Reviewer 2: It might be worth clarifying that this also used for object references.

> Examples of this constructs in C++/CLI can be found in the [official documentation][doc_cppcli].

Reviewer: ```suggestion
Examples of these constructs in C++/CLI can be found in the [official documentation][doc_cppcli].
```

> Regardless of how these constructs are used in practice their function should be clear &ndash; coordination with the [GC](#gc) to help share memory between native and managed environments.

Reviewer: ```suggestion
Regardless of how these constructs are used in practice, their function should be clear &ndash; coordination with the [GC](#gc) to help share memory between native and managed environments.
```

### Activation of the .NET runtime <a name="cppcli_activation"></a>

> Introducing .NET into an existing native application is a common scenario where C++/CLI is employed. Imagine a large native application that would like to leverage an existing .NET library. Since the applicaion is native this would imply that a .NET runtime instance is not already running in the process and the main entry point was native (for example, [`int main()`](https://en.cppreference.com/w/cpp/language/main_function)). When a C++/CLI assembly is loaded into the process, the .NET runtime must be initialized and prepared to operate in this process prior to any managed code running.

Reviewer: ```suggestion
Introducing .NET into an existing native application is a common scenario where C++/CLI is employed. Imagine a large native application that would like to leverage an existing .NET library. Since the applicaion is native, this would imply that a .NET runtime instance is not already running in the process and the main entry point is native (for example, [`int main()`](https://en.cppreference.com/w/cpp/language/main_function)). When a C++/CLI assembly is loaded into the process, the .NET runtime must be initialized and prepared to operate in this process prior to any managed code running.
```

There are are two types of .NET runtime activation for C++/CLI &ndash; .NET Framework and .NET Core 3.1/.NET 5+. These two types are mechanically similar once the .NET runtime is loaded and initialized but prior to that the details are very different.

> **.NET Framework** &ndash; The C++/CLI assembly is implicitly linked against `mscoree.lib`. The linking indicates the assembly has a dependency on the native `mscoree.dll` binary and must be loaded prior to the assembly being executed. The `mscoree.dll` binary is located globally (for example, `%SystemRoot%\System32\mscoree.dll`) on the Windows platform and facilitates loading the .NET runtime. Which version of the runtime (that is, `2.0`, `3.5`, or `4.5`+) that is loaded is influenced by additional flags passed during the compilation, a `app.config`, or global settings.

Reviewer: ```suggestion
**.NET Framework** &ndash; The C++/CLI assembly is implicitly linked against `mscoree.lib`. The linking indicates the assembly has a dependency on the native `mscoree.dll` binary and must be loaded prior to the assembly being executed. The `mscoree.dll` binary is located globally (for example, `%SystemRoot%\System32\mscoree.dll`) on the Windows platform and facilitates loading the .NET runtime. The version of the runtime (that is, `2.0`, `3.5`, or `4.5`+) that is loaded is influenced by additional flags passed during the compilation, an `app.config`, or global settings.
```

> **.NET Core 3.1** and **.NET 5+** &ndash; The C++/CLI assembly is implicitly linked against `ijwhost.lib` and thus requires `ijwhost.dll` during load. The `ijwhost.dll` binary is not necessarily globally installed and available. This makes it a dependency that must be provided in another manner (for example, update `PATH` environment variable). Once the `ijwhost.dll` is loaded, it reads an associated `<CPPCLI_ASSEMBLY_NAME>.runtimeconfig.json` file to determine which .NET runtime is needed by the assembly. If an existing .NET runtime is already loaded that version is validated it supports the needs of the requested version. Version reconciliation is non-trivial and has many options that are described in detail [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/framework-version-resolution.md).

Reviewer: > If an existing .NET runtime is already loaded that version is validated it supports the needs of the requested version.

Rephrase?

Reviewer: Do we want to link to https://github.com/dotnet/runtime/blob/main/docs/design/features/IJW-activation.md

Reviewer: > supports the needs of the requested version. Version reconciliation is non-trivial

It is also not just the version, but the existence of requested frameworks - e.g. the C++/CLI component relies `Microsoft.WindowsDesktop.App`, but the loaded runtime just has `Microsoft.NETCore.App`.

Maybe the distinction is not interesting to a lot of people, but I think it is worth clarifying. I have helped customers where they thought their component should load properly since it was targeting the same runtime version as what existed, but it turned out they were using additional frameworks.

> **.NET Core 3.1** and **.NET 5+** &ndash; The C++/CLI assembly is implicitly linked against `ijwhost.lib` and thus requires `ijwhost.dll` during load. The `ijwhost.dll` binary is not necessarily globally installed and available. This makes it a dependency that must be provided in another manner (for example, update `PATH` environment variable). Once the `ijwhost.dll` is loaded, it reads an associated `<CPPCLI_ASSEMBLY_NAME>.runtimeconfig.json` file to determine which .NET runtime is needed by the assembly. If an existing .NET runtime is already loaded, that runtime version is validated it supports the needs of the requested version and the `Microsoft.WindowsDesktop.App` framework is loaded. Version reconciliation is non-trivial and has many options that are described in detail [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/framework-version-resolution.md). The design document for this process can be found in the [`dotnet/runtime`][design_ijw_activation] repository.

Reviewer: ```suggestion
**.NET Core 3.1** and **.NET 5+** &ndash; The C++/CLI assembly is implicitly linked against `ijwhost.lib` and thus requires `ijwhost.dll` during load. The `ijwhost.dll` binary is not necessarily globally installed and available. This makes it a dependency that must be provided in another manner (for example, update `PATH` environment variable). Once the `ijwhost.dll` is loaded, it reads an associated `<CPPCLI_ASSEMBLY_NAME>.runtimeconfig.json` file to determine which .NET runtime is needed by the assembly. If an existing .NET runtime is already loaded, `ijwhost.dll` will verify that the loaded runtime supports the needs of the C++/CLI assembly. This includes validation that all requested frameworks exist and match the requested versions. Framework version reconciliation is non-trivial and has many options that are described in detail [here](https://github.com/dotnet/runtime/blob/main/docs/design/features/framework-version-resolution.md). The design document for this process can be found in the [`dotnet/runtime`][design_ijw_activation] repository.
```

> Once the specific native dependency (that is, `mscoree.dll`/`ijwhost.dll`) is loaded, all managed exports exposed to native code are "thunked". This thunk provides a level of indirection between the actual managed function and the calling native code in order to load the .NET runtime on demand. The first time a managed function is called by native code the thunk is executed and a .NET runtime loaded or the existing one confirmed compatible and adopted. All native exports are then populated with the appropriate managed function. Only one managed call must pay this initialization price since all thunks for an assembly will be updated when it is loaded into a .NET runtime. After a .NET runtime has loaded the assembly and the managed function thunks updated, the interop experience follows all the rules and principles described in the [Concepts](#concepts) section.

Reviewer: ```suggestion
Once the specific native dependency (that is, `mscoree.dll`/`ijwhost.dll`) is loaded, all managed exports exposed to native code are "thunked". This thunk provides a level of indirection between the actual managed function and the calling native code in order to load the .NET runtime on demand. The first time a managed function is called by native code, the thunk is executed and a .NET runtime loaded or the existing one confirmed to be compatible and adopted. All native exports are then populated with the appropriate managed function. Only one managed call must pay this initialization price since all thunks for an assembly will be updated when it is loaded into a .NET runtime. After a .NET runtime has loaded the assembly and the managed function thunks have been updated, the interop experience follows all the rules and principles described in the [Concepts](#concepts) section.
```
